### ISSUE TITLE
operator record Orientation

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#313]
**Modified by dietmarw on 12 Mar 2010 10:05 UTC**
Just to open the discussion, I see no way of fixing this in time in a good way for Modelica 3.2:

#### Background
For MultiBody systems the connectors contain an overdetermined orientation record with special 'equalityConstraint' to construct residuals, since an orientations has more than 3 variables (4 if quaternions, 9 if full matrix representation) but only 3 constraints.

Operator records have been recently introduced, and allow users to define operations like multiplication and addition.

#### Combined
The question is basically what would happen if Orientation was defined as an operator record:

 * How would one construct equations of type 'Orientation' when the natural equality of all elements gives too many equations?
Consider (multiplication is combined rotation, non-commutative):
```text/x-modelica
  Orientation a,b,c,d;
equation
  a*b=c*d;
// Equality of all elements gives >3 residuals,
// but only 3 residuals are really present
```
 In many cases there is no need to solve equations e.g.:
 ```
  Orientation e=a*b;
```
 I would guess the half of the solution is some kind of user-defined 'residual' in the type, similar to equalityConstraint. The second half is some way of constructing Orientation iteration variables (satisfying the class constraints).
 * If the first problem is solved could we then use an operator record Orientation in the multibody connectors and not rely on  equalityConstraint?

----

**Reported by HansOlsson on 12 Mar 2010 09:00 UTC**
Just to open the discussion, I see no way of fixing this in time in a good way for Modelica 3.2:

#### Background
For MultiBody systems the connectors contain an overdetermined orientation record with special 'equalityConstraint' to construct residuals, since an orientations has more than 3 variables (4 if quaternions, 9 if full matrix representation) but only 3 constraints.

Operator records have been recently introduced, and allow users to define operations like multiplication and addition.

#### Combined
The question is basically what would happen if Orientation was defined as an operator record:

 * How would one construct equations of type 'Orientation' when the natural equality of all elements gives too many equations?
Consider (multiplication is combined rotation, non-commutative):
[[BR]]  Orientation a,b,c,d;
equation 
  a\*b=c\*d;
 // Equality of all elements gives >3 residuals,
// but only 3 residuals are really present
In many cases there is no need to solve equations e.g.:
   Orientation e=a\*b;
[[BR]]I would guess the half of the solution is some kind of user-defined 'residual' in the type, similar to equalityConstraint. The second half is some way of constructing Orientation iteration variables (satisfying the class constraints).
 * If the first problem is solved could we then use an operator record Orientation in the multibody connectors and not rely on  equalityConstraint?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/313

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 12 Mar 2010 10:05 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 14 Mar 2010 14:01 UTC**
This is a good idea. Further sketch:

 - New overloaded operator '='.
   This is the same as the current "equalityConstrained" operator.
   As currently, if this operator has 3 residues,
   then this defines that the Rotation object has 3 independent 
   elements.

 - The tool decides whether an '=' operation is treated as
   an equality between all record elements, or whether the
   '=' function is used. For BLT-transformation this is performed 
   in the following way:
   Assignment is first done without inlining the '\*' and '=' operators.
   During this phase, every Rotation object is treated as a vector
   with 3 elements. Then, Tarjans algorithm is applied, on the still
   non-inlined operations. Afterwards, tearing selects a minimum
   number of iterations variables. The residue equations are
   then transformed to a '=' operation whereas the other 
   '=' operations are replaced by element assignments.

If this works, the overdetermined connectors are no longer needed (so Connections.branch etc.). Example:

```
  Orientation R1,R2,R3;         // 3 operator records
  input  Real phi4              // one known angle
  output Real phi1, phi2, phi3  // 3 unknown angles
equation
  R1 '=' R(phi1);
  R2 '=' R(phi2)*R1;
  R3 '=' R(phi3)*R2;
  R1 '=' R(phi4)*R3;
```

When treating the Ri as objects, they consist of 3 independent variables. Therefore we have 3\*3 (R1,R2,R3) + 3 (phi1, phi2, phi3) = 12 unknowns for 4\*3 = 12 equations. So the number of unknowns and equations is the same.

For the assignment algorithm, every R-equation is treated as a set of 3 equations. In order to thoroughly test it, the following equivalent Modelica model can be used:

```
model testRotation
  input Real phi4;
  Real phi1, phi2, phi3;
  Real R1[3], R2[3], R3[3];

  function f1
     input Real phi;
     output Real R[3];
     external "C";
  end f1;

  function f2
     input  Real R1[3];
     input  Real R2[3];
     output Real R[3];
     external "C";
  end f2;
equation
  R1 = f1(phi1);
  R2 = f2( f1(phi2), R1 );
  R3 = f2( f1(phi3), R2 );
  R1 = f2( f1(phi4), R3 );
end testRotation;
```

Running this in Dymola, gives the following flattened and processed equations (dsmodel.mof file):

```
  // Nonlinear system of equations
    // It depends on the following timevarying variables: 
    //   phi4
    // Unknowns[3]: 
    //   phi1(start = 0)
    //   phi2(start = 0)
    //   phi3(start = 0)
  algorithm // Torn part
    R1_array2.0_ := testRotation.f1(phi1);
    R1[1] := R1_array2.0_[1];
    R1[2] := R1_array2.0_[2];
    R1[3] := R1_array2.0_[3];
    R2_array2.1_ := testRotation.f2(testRotation.f1(phi2), R1);
    R2[2] := R2_array2.1_[2];
    R2[1] := R2_array2.1_[1];
    R2[3] := R2_array2.1_[3];
    R3_array2.2_ := testRotation.f2(testRotation.f1(phi3), R2);
    R3[3] := R3_array2.2_[3];
    R3[2] := R3_array2.2_[2];
    R3[1] := R3_array2.2_[1];
    R1_array2.3_ := testRotation.f2(testRotation.f1(phi4), R3);

  equation // Residual equations
    0 = R1_array2.3_[1]-R1[1];
    0 = R1_array2.3_[2]-R1[2];
    0 = R1_array2.3_[3]-R1[3];
  // End of nonlinear system of equations
```

This means, Dymola was able to make an assignment and Tarjans algorithm 
and tearing detects a loop with 3 iteration variables. These three iteration variables actually correspond to the last '=' assignment:

```
  R1 = R(phi4)*R3;
```

Therefore, the result of the analysis is, that the original equations are transformed to:

```
  R1 = R(phi1);               // 9 equations
  R2 = R(phi2)*R1;            // 9 equations
  R3 = R(phi3)*R2;            // 9 equations
  0 = R.'='(R1, R(phi4)*R3);  // 3 equations
```

which is exactly the expected equation structure, and this was derived without any connection graphs and Connections.branch(..) statement. So this is really great!

The essential remaining problem is that there is no guarantee that the residue equations all belong to the same orientation object. Intuitively, it must be possible to use some heuristics in the tearing algorithm to enforce this. So, the above strategy requires a good tearing algorithm to get efficient code. If no tearing is applied, Tarjan detects one loop with all variables and then all '=' are treated as the equalityConstrained function. So, this is still correct, but unefficient.

To summarize, formulating the orientation object with the '=' operator gives correct code, which is, however, quite inefficient. By an improved tearing algorithm it should be possible to replace most '=' operators by record assignments.

The only formal extension to Modelica is to overload the '=' operator and to add the sketched rules above.

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 15 Mar 2010 15:43 UTC**
I agree this is the main part; but in some cases it can be more complex.

Consider extending the example above with:

```
R1 = R(phi5)*R5;
```
To solve this we (assuming we don't know the inverse-formula) have to guess a value for R5 and the solve the residual equation. 

Guessing a value for R5 should only guess a value consistent with the class, i.e. the iteration variable should only have 3 degrees of freedom. However, trying to use any constructor with 3 real inputs runs into the usual problem that there does not exist a good minimal-size representation of orientations.

Thus another solution is to use orientationConstraint (as in MSL) and thus get the residuals:

```
0=R.'orientationConstraint'(R5); // 6 equations
0=R.'='(R1, R(phi5)*R5);         // 3 equations
```

Here we are allowed to guess any value of elements of R5 and the first constraint will ensure that it satisfies the class constraints (up to numeric precision); and the last equations will ensure that we pick the correct orientation among the valid ones.

More work is, of course, needed.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 21 May 2010 11:09 UTC**
Milestone ModelicaSpec3.3 deleted

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 25 Oct 2018 15:26 UTC**
Unhide tickets in preparation for move to GitHub.
(Examined tickets before.)
